src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,263 @@
+  return (*function)(x);
+}
+
+class GslQuadAdaptiveImpl {
+ public:
+  explicit GslQuadAdaptiveImpl(size_t max_intervals) noexcept;
+
+  GslQuadAdaptiveImpl() = default;
+  GslQuadAdaptiveImpl(const GslQuadAdaptiveImpl&) = delete;
+  GslQuadAdaptiveImpl& operator=(const GslQuadAdaptiveImpl&) = delete;
+  GslQuadAdaptiveImpl(GslQuadAdaptiveImpl&&) noexcept = default;
+  GslQuadAdaptiveImpl& operator=(GslQuadAdaptiveImpl&& rhs) noexcept = default;
+  ~GslQuadAdaptiveImpl() noexcept = default;
+
+  // clang-tidy: no runtime references
+  void pup(PUP::er& p) noexcept;  // NOLINT(google-runtime-references)
+
+ private:
+  struct gsl_integration_workspace_deleter {
+    void operator()(gsl_integration_workspace* workspace) const noexcept;
+  };
+
+  void initialize() noexcept;
+
+  mutable gsl_function gsl_integrand_;
+
+ protected:
+  template <typename IntegrandType>
+  gsl_function* gsl_integrand(IntegrandType&& integrand) const noexcept {
+    gsl_integrand_.function = &detail::integrand<IntegrandType>;
+    gsl_integrand_.params = &integrand;
+    return &gsl_integrand_;
+  }
+
+  size_t max_intervals_;
+  std::unique_ptr<gsl_integration_workspace, gsl_integration_workspace_deleter>
+      workspace_;
+};
+
+void check_status_code(int status_code);
+
+void disable_gsl_error_handling() noexcept;
+
+}  // namespace detail
+
+/// Defines what algorithm is used and should be chosen according
+/// to the problem.
+enum class GslIntegralType {
+  StandardGaussKronrod,            // gsl_integration_qag()
+  IntegrableSingularitiesPresent,  // gsl_integration_qags()
+  IntegrableSingularitiesKnown,    // gsl_integration_qagp()
+  InfiniteInterval,                // gsl_integration_qagi()
+  UpperBoundaryInfinite,           // gsl_integration_qagiu()
+  LowerBoundaryInfinite            // gsl_integration_qagil()
+};
+
+/*!
+ * \ingroup NumericalAlgorithmsGroup
+ * \brief A wrapper around the GSL adaptive integration procedures
+ *
+ * All templates take upper bounds to the absolute and relative error;
+ * `tolerance_abs` and `tolerance_rel(default = 0.0)`, respectively.
+ * For details on the algorithm see the GSL documentation on `gsl_integration`.
+ *
+ * Here is an example how to use this class. For the function:
+ * \snippet Test_GslQuadAdaptive.cpp integrated_function
+ * the integration should look like:
+ * \snippet Test_GslQuadAdaptive.cpp integration_example
+ */
+template <GslIntegralType TheIntegralType>
+class GslQuadAdaptive;
+
+/*!
+ * The algorithm for "StandardGaussKronrod" uses the QAG algorithm to employ an
+ * adaptive Gauss-Kronrod n-points integration rule. Its function takes a
+ * `lower_boundary` and `upper_boundary` to the integration region, an upper
+ * bound for the absolute error `tolerance_abs` and a `key`. The latter is an
+ * index to the array [15, 21, 31, 41, 51, 61], where each element denotes how
+ * many function evaluations take place in each subinterval. The GSL
```

Comment:
I don't know if you need to phrase it as "The GSL documentation mentions", this is a well-known fact about higher-order methods, so just "A higher-order rule serves..." is good